public/js/src/module/store.js

```diff
@@ -296,47 +306,49 @@ const surveyStore = {
                 }
 
-                _transformExternalDataToXmlStr( survey );
+                let update = Object.assign( {}, _serializeSurvey( survey ) );
+
+                // Note: if survey.resources = undefined/null, do not store empty array
+                // as it means there are no resources to store (and load) - they may be added later by form-cache.js
+                if ( survey.resources ) {
+                    update.resources = resourceKeys;
+                }
+
+                if ( survey.binaryDefaults ) {
+                    update.resources = binaryDefaultKeys;
+                }
 
                 // Update the existing survey
-                return server.surveys.update( {
-                    form: survey.form,
-                    model: survey.model,
-                    enketoId: survey.enketoId,
-                    hash: survey.hash,
-                    theme: survey.theme,
-                    // Note: if survey.resources = undefined/null, do not store empty array
-                    // as it means there are no resources to store (and load) - they may be added later by form-cache.js
-                    resources: survey.resources ? resourceKeys : undefined,
-                    binaryDefaults: survey.binaryDefaults ? binaryDefaultKeys : undefined,
-                    maxSize: survey.maxSize,
-                    externalData: survey.externalData,
-                    branding: survey.branding
-                } );
+                return server.surveys.update( update );
             } )
-            .then( () => {
+            .then( _firstItemOnly )
+            .then( result => {
                 const tasks = [];
                 // Add new or update existing resources, combining binaryDefaults and form resources
                 ( survey.resources || [] ).concat( survey.binaryDefaults || [] ).forEach( file => {
                     tasks.push( surveyStore.resource.update( survey.enketoId, file ) );
                 } );
                 // Remove obsolete resources
                 obsoleteResources.concat( obsoleteBinaryDefaults ).forEach( key => {
-                    tasks.push( surveyStore.resource.remove( survey.enketoId, key ) );
+                    tasks.push( surveyStore.resource.remove( survey.enketoId, key ).then( () => null ) );
                 } );
 
                 // Execution
                 return Promise.all( tasks )
-                    .then( () => // resolving with original survey (not the array returned by server.surveys.update)
-                        survey )
-                    .then( _transformExternalDataToXmlDoc );
+                    .then( resources => {
+                        if ( survey.resources ) {
+                            result.resources = resources.filter( resource => resource != null );
+                        }
+
+                        return _deserializeSurvey( result );
+                    } );
```

Comment:
This was also about minimizing the difference between create/update, but also and more importantly about being able to validate that the actual update produces the expected result, where returning the incoming parameter might mask problems.